unittests/test-contracts/savanna/common/savanna.hpp

```diff
@@ -292,53 +296,38 @@ namespace savanna {
       //finalizer_policy_generation for this block
       uint32_t finalizer_policy_generation;
 
-      //if a valid qc is obtained over the digest generated from hashing the content of this message, it is a proof of finality for the block number recorded here
-      uint32_t final_on_strong_qc_block_num = 0;
-
       //if a new finalizer policy is promoted to last pending status, it could (but is not guaranteed to) become active.
       //Therefore, we provide a mechanism to include finalizer policies into a proof of finality.
       //This allows the contract to obtain knowledge about them and to record them in its internal state.
       std::optional<finalizer_policy_input> new_finalizer_policy;
-
-      std::optional<checksum256> reversible_blocks_mroot;
-
       std::optional<uint32_t> last_pending_finalizer_policy_start_num;
+      //std::optional<checksum256> level_3_commitments_digest;
 
       //if a finalizer policy is present, witness_hash should be the base_digest. Otherwise, witness_hash should be the static_data_digest
       checksum256 witness_hash;
 
-      //finality merkle root for final_on_strong_qc_block_num
+      //finality merkle root
       checksum256 finality_mroot;
       
       //returns hash of digest of new_finalizer_policy + witness_hash if new_finalizer_policy is present, otherwise returns witness_hash
       checksum256 resolve_witness() const {
-         if (new_finalizer_policy.has_value() && reversible_blocks_mroot.has_value()){
+         if (new_finalizer_policy.has_value() 
+            && last_pending_finalizer_policy_start_num.has_value()
+            && witness_hash!=checksum256()){
+
             checksum256 policy_digest = new_finalizer_policy.value().digest();
             
-            uint32_t resolved_last_pending_finalizer_policy_start_num = 0;
-
-            if (last_pending_finalizer_policy_start_num.has_value()) 
-               resolved_last_pending_finalizer_policy_start_num = last_pending_finalizer_policy_start_num.value();
-
-            auto l3_packed = eosio::pack(level_3_commitments_t{
-               .reversible_blocks_mroot  = reversible_blocks_mroot.value() , 
-               .base_digest = witness_hash
-            });
-
-            checksum256 l3_digest = sha256(l3_packed.data(), l3_packed.size());
-
             auto l2_packed = eosio::pack(level_2_commitments_t{
                .last_pending_fin_pol_digest  = policy_digest, 
-               .last_pending_fin_pol_start_num = resolved_last_pending_finalizer_policy_start_num,
-               .l3_commitments_digest = l3_digest
+               .last_pending_fin_pol_start_num =  last_pending_finalizer_policy_start_num.value(),
+               .l3_commitments_digest = witness_hash

```

Comment:
Is `witness_hash` the same as l3_commitments?

Reply:
Sometimes.

I added a comment to explain it.